src/main/java/com/microsoft/sqlserver/jdbc/ConfigurableRetryLogic.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Microsoft JDBC Driver for SQL Server Copyright(c) Microsoft Corporation All rights reserved. This program is made
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+
+package com.microsoft.sqlserver.jdbc;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.text.MessageFormat;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+
+/**
+ * Allows configurable statement retry through the use of the 'retryExec' connection property. Each rule read in is
+ * converted to ConfigRetryRule objects, which are stored and referenced during statement retry.
+ *
+ */
+public class ConfigurableRetryLogic {
+    private final static int INTERVAL_BETWEEN_READS_IN_MS = 30000;
+    private final static String DEFAULT_PROPS_FILE = "mssql-jdbc.properties";
+    private static final java.util.logging.Logger CONFIGURABLE_RETRY_LOGGER = java.util.logging.Logger
+            .getLogger("com.microsoft.sqlserver.jdbc.ConfigurableRetryLogic");
+    private static ConfigurableRetryLogic singleInstance = null;

```

Comment:
Consider making singeInstance singleton variable volatile. It looks like you are using the Double Checked Locking Idiom to perform the synchronization (i.e. check null, lock, and recheck null). However this does not work in Java. Please check this blog describing the issue in both C++ and Java:
https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

Here is a 2nd blog describing the issue and why volatile fixes the problem:
https://stackoverflow.com/questions/11639746/what-is-the-point-of-making-the-singleton-instance-volatile-while-using-double-l

Reply:
Good comment, but the `singleInstance` is actually thread-safe because its a static initializer, so I can just remove the locks altogether.

Reply:
It is statically initialized to NULL so there is no ConfigurableRetryLogic object instance. We are using lazy construction in getInstance() to create the object at dynamically at run time.

Reply:
by load time I mean when the java class is loaded into memory and all the appropriate static methods have been invoked but before a client thread is given access to the classes in the JAR. In this case it would be either a thread from the calling application or a thread that we created at run time. This is why I am saying that we are using lazy construction. It means that the ConfigurableRetryLogic instance is not constructed by Java which is thread safe but by a client thread or a thread spawned by a client thread at run time. Just because the getInstance() method is static, it does not mean it is implicitly thread safe. It only means that there is one instance and that multiple threads can and will access that static method